api/app/utils/cffdrs.py

```diff
@@ -296,11 +401,120 @@ def get_ffmc_for_target_hfi(fuel_type: str,
         experimental_hfi = head_fire_intensity(fuel_type,
                                                percentage_conifer,
                                                percentage_dead_balsam_fir,
-                                               bui, experimental_ffmc, ros, cfb, cfl, experimental_sfc)
+                                               bui, experimental_ffmc, experimental_ros,
+                                               cfb, cfl, experimental_sfc)
         error_hfi = (target_hfi - experimental_hfi) / target_hfi
 
     return (experimental_ffmc, experimental_hfi)
+
+
+def get_critical_hours_start(critical_ffmc: float, solar_noon_ffmc: float, temperature: float,
+                             relative_humidity: float, wind_speed: float, precip: float):
+    """ Returns the hour of day (on 24H clock) at which the hourly FFMC crosses the
+    threshold of critical_ffmc.
+    Returns None if the hourly FFMC never reaches critical_ffmc.
+    """
+    if solar_noon_ffmc >= critical_ffmc:
+        logger.info('Solar noon FFMC >= critical FFMC')
+        # go back in time in increments of 0.5 hours
+        clock_time = 13-0.5  # start from solar noon - 0.5 hours
+        while get_hourly_ffmc_on_diurnal_curve(solar_noon_ffmc, clock_time, temperature,
+                                               relative_humidity, wind_speed, precip) >= critical_ffmc:
+            clock_time -= 0.5
+            if clock_time == -0.5:
+                break
+        # add back the half hour that caused FFMC to drop below critical_ffmc (or that pushed time below 0.0)
+        clock_time += 0.5
+        logger.info('%s', clock_time)
+        return clock_time
+
+    logger.info('Solar noon FFMC %s < critical FFMC %s', solar_noon_ffmc, critical_ffmc)
+    # go forward in time in increments of 0.5 hours
+    clock_time = 13 + 0.5  # start from solar noon + 0.5 hours
+    while get_hourly_ffmc_on_diurnal_curve(solar_noon_ffmc, clock_time, temperature,
+                                           relative_humidity, wind_speed, precip) < critical_ffmc:
+        logger.info('Clock time %s has HFFMC %s', clock_time, get_hourly_ffmc_on_diurnal_curve(
+            solar_noon_ffmc, clock_time, temperature, relative_humidity, wind_speed, precip))
+        clock_time += 0.5
+        if clock_time == 24.0:
+            return None
+    return clock_time
+
+
+def get_critical_hours_end(critical_ffmc: float, solar_noon_ffmc: float, critical_hour_start: float,
+                           temperature: float, relative_humidity: float, wind_speed: float, precip: float):
+    """ Returns the hour of day (on 24H clock) at which the hourly FFMC drops below
+    the threshold of critical_ffmc.
+    Should only be called if critical_hour_start is not None.
+    If diurnally-adjusted FFMC never drops below critical_ffmc in the day, will return 23.5 (11:30 pm).
+    """
+    assert critical_hour_start is not None
```

Comment:
Is this assert here just for testing? Can it be removed?

Reply:
it's for testing/troubleshooting purposes - will be removed in next PR once we get diurnal FFMC calcs working correctly